solutions/gold/izho-17-HardRoute.mdx

```diff
@@ -0,0 +1,149 @@
+---
+id: izho-17-HardRoute
+source: IZhO 2017
+title: Hard Route
+author: Justin Ji
+---
+
+## Explanation
+
+Assume there is some hard route that goes from vertex $u$ to vertex $v$.
+Let the node that the path from $u$ to $v$ and the furthest node from the hard route meet be node $x$.
+Use rerooting DP to calculate the hardest route and the number of such hardest routes for every node $x$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+## Implementation
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+const int N = 5e5 + 1;
+int max_length[N], path_count[N];
+vector<int> adj[N];
+ll max_hardness = 0, hardest_path_count = 1;
+void dfs(int u, int p) {
+	/**
+	 * Calculates the longest path from vertex u,
+	 * and the number of such paths.
+	 */
+	max_length[u] = 0;
+	path_count[u] = 1;
+	for (int v : adj[u])
+		if (v != p) {
+			dfs(v, u);
+			if (max_length[u] < max_length[v] + 1) {
+				max_length[u] = max_length[v] + 1;
+				path_count[u] = path_count[v];
+			} else if (max_length[v] + 1 == max_length[u]) {
+				path_count[u] += path_count[v];
+			}
+		}
+}
+void dfs2(int u, int p, ll parDist, ll parCnt) {
```

Comment:
uuuhh i thought i asked for `std::function`? sorry for the confusion
or does it not run in time i forgor

Reply:
nah std::function would probably run in time, I just felt like doing function<void(int, int, ll, ll)> is kind of ugly

Reply:
can you uh please change it to std function
or do you just want me to do it for you

Reply:
lmao aight ill change it, gimme a sec